### PR TITLE
request_id: Add option to always set request id in response

### LINKED ISSUE #10807: request_id: Allow always setting x-request-id header in response
*Description*:
Currently non-configurable Envoy behavior is to return `x-request-id` header in response iff the tracing is forced.
It would be nice to have a configuration option to allow always returning the header in response.

It is especially useful for cross-correlating client-side debug logs with access logs on server side.

### PR DESCRIPTION
**Pull request by euroelessar**
Description: Add configuration option to return `x-request-id` in all cases, even if tracing is not forced.
Risk Level: LOW (disabled by default)
Testing: unit tests
Docs Changes: proto documentation
Release Notes: updated
Fixes #10807

### CODE REVIEW

**Comment by euroelessar** [comment]
This is a followup for one of the discussion branches in #10429.

**Comment by euroelessar** [comment]
clang-tidy looks like false positive:
```
2020-04-16T09:26:08.1123546Z source/common/http/conn_manager_config.h:14:10: error: 'common/stats/symbol_table_impl.h' file not found [clang-diagnostic-error]
2020-04-16T09:26:08.1124954Z #include "common/stats/symbol_table_impl.h"
2020-04-16T09:26:08.1125360Z          ^
2020-04-16T09:26:08.1129473Z 
2020-04-16T09:26:08.1129806Z 59922 warnings and 1 error generated.
2020-04-16T09:26:08.1130302Z Error while processing /source/source/common/http/conn_manager_config.h.
```

Same for coverage:
```
[----------] 1 test from EvironmentCredentialsProviderTest
[ RUN      ] EvironmentCredentialsProviderTest.MissingAccessKeyId
test/extensions/common/aws/credentials_provider_impl_test.cc:51: Failure
Value of: credentials.accessKeyId().has_value()
  Actual: true
Expected: false
Stack trace:
  0x8c2e26f: Envoy::Extensions::Common::Aws::EvironmentCredentialsProviderTest_MissingAccessKeyId_Test::TestBody()
  0x121efa46: testing::internal::HandleSehExceptionsInMethodIfSupported<>()
  0x121dc011: testing::internal::HandleExceptionsInMethodIfSupported<>()
  0x121c68e2: testing::Test::Run()
  0x121c73e8: testing::TestInfo::Run()
... Google Test internal frames ...

test/extensions/common/aws/credentials_provider_impl_test.cc:52: Failure
Value of: credentials.secretAccessKey().has_value()
  Actual: true
Expected: false
Stack trace:
  0x8c2e43f: Envoy::Extensions::Common::Aws::EvironmentCredentialsProviderTest_MissingAccessKeyId_Test::TestBody()
  0x121efa46: testing::internal::HandleSehExceptionsInMethodIfSupported<>()
  0x121dc011: testing::internal::HandleExceptionsInMethodIfSupported<>()
  0x121c68e2: testing::Test::Run()
  0x121c73e8: testing::TestInfo::Run()
... Google Test internal frames ...

[  FAILED  ] EvironmentCredentialsProviderTest.MissingAccessKeyId (1420 ms)
```

File: api/envoy/extensions/filters/network/http_connection_manager/v4alpha/http_connection_manager.proto [hunk 1]
```diff
@@ -433,6 +433,11 @@ message HttpConnectionManager {
   // is the current Envoy behaviour. This defaults to false.
   bool preserve_external_request_id = 32;
 
+  // If set, Envoy will always set :ref:`x-request-id <config_http_conn_man_headers_x-request-id>` header in response.
+  // If this is false or not set, the header is returned in responses only in case of forced tracing.
+  // This defaults to false.
```
Comment (htuch) [inline api/envoy/extensions/filters/network/http_connection_manager/v4alpha/http_connection_manager.proto:438]: No need to specify a default for proto3 scalars, this is implied.

File: api/envoy/extensions/filters/network/http_connection_manager/v4alpha/http_connection_manager.proto [hunk 2]
```diff
@@ -433,6 +433,11 @@ message HttpConnectionManager {
   // is the current Envoy behaviour. This defaults to false.
   bool preserve_external_request_id = 32;
 
+  // If set, Envoy will always set :ref:`x-request-id <config_http_conn_man_headers_x-request-id>` header in response.
+  // If this is false or not set, the header is returned in responses only in case of forced tracing.
```
Comment (htuch) [inline api/envoy/extensions/filters/network/http_connection_manager/v4alpha/http_connection_manager.proto:437]: Can you make "forced tracing" a bit clearer for the reader?

**Comment by euroelessar** [comment]
I'm not sure what happened with circlici, it tells about compilation errors in a code which does not exist for lines which are not there in the last version of pr.
I'll just rebase to hopefully make it go away.

**Comment by yanavlasov** [comment]
#10860 should fix your clang-tidy error

**Comment by euroelessar** [comment]
@yanavlasov clang-tidy is still broken:
```
source/common/http/conn_manager_config.h:14:10: error: 'common/stats/symbol_table_impl.h' file not found [clang-diagnostic-error]
#include "common/stats/symbol_table_impl.h"
```

**Comment by mattklein123** [comment]
Merge current master. You don't have the fix.

/wait

**Comment by euroelessar** [comment]
Oh, merged. Misread the comment.

**Comment by euroelessar** [comment]
@htuch @mattklein123 All tests did pass, PTAL

**Pull request merged**